code/gateway/gatewaylib.q

```diff
@@ -92,3 +92,199 @@ getserverscross:{[req;att;besteffort]
  /- return the parameters which should be queried for
  (key s)!distinct each' flip each util[w]`found
  }
+
+// Input a table of ([]tablename;starttime;endtime;instruments;procs)
+// Or dict of `tablename`starttime`endtime`instruments`procs!(tablename;starttime;endtime;instruments;procs)
+// Returns a dict of serverid(s) that should be queried to cover all that data
+getservers:{[dict]
+    // Check if input is a table, get the dict
+    if[98h~type dict;
+        if[1<count dict;.z.s each dict]; /use recursive calls for each dict(row)
+        if[1=count dict;dict:first dict]]; 
+
+    // Check required keys
+    req:`tablename`starttime`endtime;
+    if[not all req in k:key dict;
+        '"Provide a dictionary with required keys: `tablename`starttime`endtime"];
+
+    // Check if correct keys
+    if[not all k in`tablename`starttime`endtime`instruments`procs;
+        '"Provide a dictionary with keys of only: `tablename`starttime`endtime`instruments`procs"];
+
+    // Checktype
+    dict:.checkinputs.checkinputs dict;
+
+    // Get the Procs in a (nested) list of serverid(s)
+    dict[`procs]:attributesrouting[dict;part:partdict dict];
+
+    // Adjust queries based on relevant data and stripe
+    d:adjustqueriesoverlap[dict;part];
+    servers:adjustqueriesstripe[dict;d];
+	:key[servers]!(`serverid`attributes`checksperformed`optimhdb)_value servers;
+    };
+
+// Dynamic routing finds all processes with relevant data 
+attributesrouting:{[options;procdict]
```

Comment:
this only does routing based on times, not the instrument striping stuff, is that right?

Reply:
that's right yes, the instruments routing comes later in adjustqueriesstripe (also in gatewaylib.q)